### DISCUSSION TITLE
Showing opaque values using Sexp

### DISCUSSION THREAD

**Discussion by Zeta611** [hackwaly/ocamlearlybird#51] [Q&A]
Hi, I'm wondering if I can use ocamlearlybird to show opaque values using sexp.

I'm using ocamlearlybird in VSCode.

I made some custom modules implementing `to_sexp : t -> Sexp.t` so that I can inspect opaque values using s-expressions.
Is there a way to inspect them using my custom `to_sexp`?
By "opaque values", I mean:
<img width="223" alt="Screenshot 2023-08-05 at 11 29 07" src="https://github.com/hackwaly/ocamlearlybird/assets/9553691/c80c9445-c942-46c7-8e7c-07f5893c2501">

Fully-automatic solution would be cool, but I would still be very happy if I can manually invoke this method for variables in scope at a breakpoint.
Is there currently a way to do this?

Thank you for creating and maintaining this amazing tool!

**Comment by sim642** [comment]
I recall it being mentioned somewhere that it could be possible to at least register and use `Format` printers for types like the toplevel does because this is also all in bytecode. However, I'm not familiar with that somewhat magic system and haven't looked into it. `ocamldebug` itself has that: https://v2.ocaml.org/manual/debugger.html#ss:debugger-printers.

A similar system might work for registering more structured type-based output as well, either from s-exps or JSON.

**Reply by Zeta611** [reply]
Thank you for providing the pointer to the relevant doc for `ocamldebug`.
